src/libraries/Common/src/System/Net/Http/Http3/Helpers/VariableLengthIntegerHelper.cs

```diff
@@ -0,0 +1,222 @@
+
+        public static bool TryRead(ReadOnlySpan<byte> buffer, out long value, out int bytesRead)
+        {
+            if (buffer.Length == 0)
+            {
+                goto needMore;
+            }
+
+            // The first two bits of the first byte represent the length of the
+            // variable length integer
+            // 00 = length 1
+            // 01 = length 2
+            // 10 = length 4
+            // 11 = length 8
+
+            byte firstByte = buffer[0];
+
+            switch (firstByte & LengthMask)
+            {
+                case LengthOneByte:
+                    value = firstByte;
+                    bytesRead = 1;
+                    return true;
+                case LengthTwoByte:
+                    if (buffer.Length < 2)
+                    {
+                        goto needMore;
+                    }
+                    value = BinaryPrimitives.ReadUInt16BigEndian(buffer) - TwoByteSubtract;
+                    bytesRead = 2;
+                    return true;
+                case LengthFourByte:
+                    if (buffer.Length < 4)
+                    {
+                        goto needMore;
+                    }
+                    value = BinaryPrimitives.ReadUInt32BigEndian(buffer) - FourByteSubtract;
+                    bytesRead = 4;
+                    return true;
+                default:
+                    Debug.Assert((firstByte & LengthMask) == LengthEightByte);
+
+                    if (buffer.Length < 8)
+                    {
+                        goto needMore;
+                    }
+                    value = (long)(BinaryPrimitives.ReadUInt64BigEndian(buffer) - EightByteSubtract);
+                    bytesRead = 4;
+                    return true;
+            }
+
+        needMore:
+            value = 0;
+            bytesRead = 0;
+            return false;
+        }
+
+        public static bool TryRead(ref SequenceReader<byte> reader, out long value)
+        {
+            ReadOnlySpan<byte> span = reader.UnreadSpan;
+
+            if (span.Length >= 8)
+            {
+                // Hot path: call span-based read immediately.
+                if (TryRead(span, out value, out int bytesRead))
+                {
+                    reader.Advance(bytesRead);
+                    return true;
+                }
+                else
+                {
+                    return false;
+                }
+            }
+
+            // Cold path: copy to a temporary buffer before calling span-based read.
+            return TryReadSlow(ref reader, out value);
+
+            static bool TryReadSlow(ref SequenceReader<byte> reader, out long value)
+            {
+                ReadOnlySpan<byte> span = reader.CurrentSpan;
+
+                if (!reader.TryPeek(out byte firstByte))
+                {
+                    value = 0;
+                    return false;
+                }
+
+                int length =
```

Comment:
This calculation seems overboard. I think it would just be simpiler to always stackalloc 8 bytes, copy, and then call TryRead

Reply:
`TryCopyTo` would fail if we asked for 8 bytes but there were only 4 available.